### PR TITLE
Update contributing.rst

### PR DESCRIPTION
**Pull request by kbx81**
## Description:

Various updates, particularly around PR submission. Lots of other tweaks & copy fixes/corrections.

**Related issue (if applicable):** fixes <link to issue>

**Pull request in [esphome](https://github.com/esphome/esphome) with YAML changes (if applicable):** esphome/esphome#<esphome PR number goes here>

## Checklist:

  - [ ] I am merging into `next` because this is new documentation that has a matching pull-request in [esphome](https://github.com/esphome/esphome) as linked above.  
    or
  - [x] I am merging into `current` because this is a fix, change and/or adjustment in the current documentation and is not for a new component or feature.

  - [ ] Link added in `/index.rst` when creating new documents for new components or cookbook.

### CODE REVIEW

File: guides/contributing.rst [hunk 1]
```diff
@@ -517,31 +617,32 @@ called ``CONFIG_SCHEMA``. An example of such a schema is shown below:
...
-that converts the user input to C++ code.
+As you may know, ESPHome "converts" the user's YAML configuration into C++ code (you can see the generated code under
+``<NODE_NAME>/src/main.cpp``). Each component must define its own ``to_code`` method that "converts" the user input to
+C++ code.
 
-This method is also automatically loaded and invoked by the ESPHome core. An example of
-such a method can be seen below:
+This method is also automatically loaded and invoked by the ESPHome core. Here's an example of such a method:
 
 .. code-block:: python
```
Comment (jesserockz) [inline guides/contributing.rst:698]: Can we update this block to use async/await and also update the text below to remove all references to yield and use await.

File: guides/contributing.rst [hunk 2]
```diff
@@ -476,34 +575,35 @@ Suppose the user types in the following:
     sensor:
       - platform: hello2
 
-In both cases, ESPHome will automatically look for corresponding entries in the "components"
-folder and find the directory with the given name. For example the first entry will make ESPHome
-look at the ``esphome/components/hello1/__init__.py`` file and the second entry will result in
+In both cases, ESPHome will automatically look for corresponding entries in the "components" folder and find the
+directory with the given name. In this example, the first entry causes ESPHome to look for the
+``esphome/components/hello1/__init__.py`` file and the second entry tells ESPHome to look for
 ``esphome/components/hello2/sensor.py``.
```
Comment (jesserockz) [inline guides/contributing.rst]:
```suggestion
``esphome/components/hello2/sensor.py`` or ``esphome/components/hello2/sensor/__init__.py``.
```

File: guides/contributing.rst [hunk 3]
```diff
@@ -592,83 +693,83 @@ if communication works (if not, it should call ``mark_failed()``).
...
+:apiclass:`PollingComponent`). These methods should retrieve the latest data from your component and publish them with
+the provided methods.
 
-Finally, your component should have a ``dump_config`` method that prints the user configuration.
+.. note::
+
+    **Code in** ``loop()``, ``update()`` **and** ``setup()`` **must not block**. Specifically, methods like ``delay()``
+    should be avoided and **delays longer than 10 ms are not permitted**. The reason for this is that ESPHome uses a
+    single-threaded loop for all components -- if your component blocks, it will delay the whole loop, negatively
+    impacting other components. This can result in a variety of problems such as network connections being lost.
```
Comment (jesserockz) [inline guides/contributing.rst]: Maybe here give reference to `this->set_timeout` or using a state-machine to handle "waiting" in loop for the state to become what is needed for the next stage?

File: guides/contributing.rst [hunk 4]
```diff
@@ -592,83 +693,83 @@ if communication works (if not, it should call ``mark_failed()``).
...
-- ``MULTI_CONF``: Mark this component to accept an array of configurations. If this is an
-  integer instead of a boolean, validation will only permit the given number of entries.
-- ``CONFLICTS_WITH``: Mark a list of components as conflicting with this component. If the user
-  has one of them in the config, a validation error will be generated.
-
+- ``MULTI_CONF``: Mark this component to accept an array of configurations. If this is an integer instead of a boolean,
+  validation will only permit the given number of entries.
+- ``CONFLICTS_WITH``: Mark a list of components as conflicting with this component. If the user has one of them in
+  their config, a validation error will be generated.
 - ``ESP_PLATFORMS``: Provide a list of allowed ESP types this component works with.
```
Comment (jesserockz) [inline guides/contributing.rst]: Not used anymore
```suggestion
```

**Review by jesserockz — CHANGES_REQUESTED** [review]

File: guides/contributing.rst (hunk 1 above)
Comment (kbx81) [reply guides/contributing.rst:698]: See how this looks: [`33d4838` (#4188)](https://github.com/esphome/esphome-docs/pull/4188/commits/33d48387bcba798c6ea072ee5db0a5ebed32f61d)

File: guides/contributing.rst (hunk 3 above)
Comment (kbx81) [reply guides/contributing.rst]: See how this looks: [`33d4838` (#4188)](https://github.com/esphome/esphome-docs/pull/4188/commits/33d48387bcba798c6ea072ee5db0a5ebed32f61d)

File: guides/contributing.rst [hunk 5]
```diff
@@ -112,8 +111,8 @@ Build
 
         docker run --rm -v "${PWD}/":/workspaces/esphome-docs -p 8000:8000 -it ghcr.io/esphome/esphome-docs
 
-    With ``PWD`` referring to the root of the ``esphome-docs`` git repository. Then go to ``<CONTAINER_IP>:8000`` in
-    your browser.
+    ...with ``PWD`` referring to the root of the ``esphome-docs`` git repository. Then, to see the preview, go to
+    ``<CONTAINER_IP>:8000`` in your browser.
```
Comment (jesserockz) [inline guides/contributing.rst]: Containers dont have IPs 
```suggestion
    ``<HOST_IP>:8000`` in your browser.
```

Comment (kbx81) [reply guides/contributing.rst]: 🤦🏻 Good catch, haha

**Review by jesserockz — DISMISSED** [review]

**Review by jesserockz — APPROVED** [review]

File: guides/contributing.rst [hunk 6]
```diff
@@ -697,14 +747,28 @@ The next method that will be called with your component is ``loop()`` (or ``upda
 :apiclass:`PollingComponent`). These methods should retrieve the latest data from your component and publish them with
 the provided methods.
 
-.. note::
+Finally, your component must have a ``dump_config`` method that prints the complete user configuration.
```
Comment (nielsnl68) [inline guides/contributing.rst:751]: when **`dump_config`** is an required method to be override. would it not good idea to change up the Component class by changing "`Virtual dump_config() {}`" into "`Virtual dump_config() = 0;`"?

**Pull request merged**